### DISCUSSION TITLE
Documentation and debugging for developers

### DISCUSSION THREAD

**Discussion by PedroLSF** [mortbopet/Ripes#351] [Q&A]
Dear @mortbopet and developers,

I have some doubts at different points, I will describe them here.
Firstly, I used Docker to try to explore the project, so far so good.
![image](https://github.com/mortbopet/Ripes/assets/85000470/47d57ae0-ae8b-4c15-aafc-c5767e317792)

I basically changed the endpoint to use CLI mode
![image](https://github.com/mortbopet/Ripes/assets/85000470/dfedeb99-5971-453a-8392-3b1cc9873e13)

But I went looking for read syscall and couldn't find it. (I used the value 5/63 to read integers, but it doesn't work)
![image](https://github.com/mortbopet/Ripes/assets/85000470/444cfe8c-17fd-467e-9c6c-fa4fc58a0734)

This was the first point that raised my doubts, is there somewhere with this documented or is there documentation for developers to fully understand the project?

Finally, my last question is to know how to debug, as I didn't find the documentation, I wanted to debug or put some prints to see the data moving between the methods and variables.
https://github.com/mortbopet/Ripes/blob/0faf41b669a93a1944707cd7d111a5e9241425fe/src/syscall/systemio.h#L458

For example, in the line above I would like to see the value that is arriving at &data through a print or debug, can you help me at this point too?

Thank you very much in advance.

**Comment by federicovilla55** [comment]
As described in the [ecalls.md](https://github.com/mortbopet/Ripes/blob/master/docs/ecalls.md) file in the docs folder, all supported syscall are described in the `Help-> System calls` menù.

Trying to execute the following code in the GUI version, the input is read and in the `x10` (`a0`) register it is returned the number of read bytes.

```assembly
addi x10 x0 0
addi x17 x0 63
addi x12 x0, 255
ecall
addi x17 x0 10
ecall
```

If you want to observe the value arriving at `&data`, one approach is to include a [`qDebug()`](https://doc.qt.io/qt-6/qdebug.html) statement for each variable you wish to print, like so:

```C
void putStdInData(const QByteArray &data) {
    qDebug() << data;
    ...
}
```

Then build your local version of Ripes as described [here](https://github.com/mortbopet/Ripes?tab=readme-ov-file#building).

**Comment by PedroLSF** [comment]
@federicovilla55 Thanks for the answer!
I understood about the ecalls, I'll take a look.

Regarding qDebug(), I still haven't been able to handle it correctly.
![image](https://github.com/mortbopet/Ripes/assets/85000470/b0277be8-c851-4a23-ad42-fbf2aec9f206)

I'm inside my fork, I used qDebug() as you mentioned, I committed and did docker build again and finally I ran it with docker run, but without success when using the console, as shown below:
![image](https://github.com/mortbopet/Ripes/assets/85000470/1f837dff-4c8f-470a-ace1-d562b03b18f6)

So, since it didn't work, does that mean I can't use docker to use the console and I have to build Ripes manually? (I would very much like to stay with docker)

Note: It is worth mentioning that I am only using it in CLI mode, as I need to connect with a type of Online Judge

**Reply by federicovilla55** [reply]
@PedroLSF 
Unfortunately in the current version there's no option to handle system input from CLI.
I've created an issue explaining the problem and proposing a solution to handle stdin from CLI. [Issue #353](https://github.com/mortbopet/Ripes/issues/353)

The problem you found wasn't related to Docker, just to the current implementation of Ripes. 

If you modify _puStdInData()_ you modify a method called by the console, which is created only in the GUI mode.

Hope this helps.

**Reply by PedroLSF** [reply]
@federicovilla55 
Yes, I understand that.
I wanted to see something about this [Issue](https://github.com/mortbopet/Ripes/issues/353), but I'm not that familiar with the project and it was very difficult to understand or develop something without being able to debug or see how the data is moving within the code.
I would be very happy to help or contribute something.
